### ISSUE TITLE
What Needs Changed to Implement Individual Item Loading

### ISSUE THREAD

**Issue by rorycor** [cvdlab/react-planner#149]
I created an issue about this in the past which was more of a generic asking if it could be done but I figured I'd open this issue to at the very least get pointed in the right direction.

I've been trying to get my items to load individually so that items that use the same planner element but with different values determining their textures have the correct textures when the scene is loaded, currently the last item determines the appearance of the rest of the items as it's the last item to be loaded going into the mtl loader.

I've done some testing with breakpoints on this issue to see if I could figure out what needs changed myself but the program seems to loop through a few functions before reaching the mtl loader which makes it hard to establish the individual changes that need to be made to get the system to work differently.

I'm guessing the changes that have to be made either have to be made to the render3D function of the planner-element, the load-obj util or the mtl loader but what needs to be changed hasn't jumped out at me nor has any tinkering I've done myself yielded any results. All I've manged to establish is that each item does seem to be looped through individually but the mtl loader is only run once after the last item is loaded.

I did look into the three.js documentation on their mtl loader as mentioned in the previous issue but it is more for setting it up and didn't really help me about what I'd need to change in order to run it for every item.

If anybody has any clue as to what I need to look into doing I'd appreciate the input, I've tried asking some other forms as I didn't want to hassle here about the same issue but so far I've had no luck getting an answer.

**Comment by nasriachref** [comment]
if i understand correctly you have multiple items sharing the same planner element and you load item's material with mtl-loader .
if so the problem is that mtl-loader is an asynchronous task which mean the loading of the object and the loading of the materials are somewhat not linked(in time) to eachother and in your case maybe it overwrite your values before the texture is fully loaded.
my question is that mtl-loader is it realy necessary ?  you could  write write a function in the element code that create material (using three js StandardMaterials) and apply texture to your item given a value input and call it in the render3D function.
best of  luck to you

**Comment by rorycor** [comment]
Thanks again for the input, it's honestly great getting a good explanation of how the system ties together. 

The main reason I went straight for re-tooling the mtl-loader is because I happened to pick the sofa as a baseline for making my own items and by chance and it ended up being perfect for what I was doing with Blender.

You're 100% about my values getting overwritten, I was passing through a variable from the planner-element to the mtl-loader to determine if certain sides should be textured:

if (mat.name == 'dpGableLeft' && dpNumber1 != 1) {
break;
}
if (mat.name == 'dpGableRight' && dpNumber2 != 1) {
break;
}

however since the mtl-loader wasn't being called between items the variables were just being taken from the last instance.

I actually just got a workaround working just there by creating a function to pass through the variables to an array which is then used instead:

if (mat.name == 'dpGableLeft' && dpArray1[i] != 1) {
break;
}
if (mat.name == 'dpGableRight' && dpArray2[i] != 1) {
break;
}

If this starts giving me bother in the future the element code will be a nice fallback.
